pkg/planner/core/partition_prune.go

```diff
@@ -38,6 +38,48 @@ func PartitionPruning(ctx base.PlanContext, tbl table.PartitionedTable, conds []
 			return nil, err
 		}
 		ret := s.convertToIntSlice(rangeOr, pi, partitionNames)
+		if pi.CanHaveOverlappingDroppingPartition() {
+			if len(ret) == 1 && ret[0] == FullRange {
+				ret = make([]int, 0, len(pi.Definitions))
+				for i := range pi.Definitions {
+					ret = append(ret, i)
+				}
+			}
+			newRet := make([]int, 0, len(ret))
+			for i := range ret {
+				idx := pi.GetOverlappingDroppingPartitionIdx(ret[i])
+				if idx == -1 {
+					// dropped without overlapping partition, skip it
+					continue
+				}
+				if idx == ret[i] {
+					// non-dropped partition
+					newRet = append(newRet, idx)
+					continue
+				}
+				// partition being dropped, remove the consecutive range of dropping partitions
+				// and add the overlapping partition.
+				end := i + 1
+				for ; end < len(ret) && ret[end] < idx; end++ {
+					continue
+				}
+				// add the overlapping partition, if not already included
+				if end >= len(ret) || ret[end] != idx {
+					// It must also match partitionNames if explicitly given
+					if len(partitionNames) == 0 || s.findByName(partitionNames, pi.Definitions[idx].Name.L) {
+						newRet = append(newRet, idx)
+					}
+				}
+				if end < len(ret) {
+					newRet = append(newRet, ret[end:]...)
+				}
+				break
+			}
+			if len(newRet) == len(pi.Definitions) {
+				return []int{FullRange}, nil
+			}
+			return newRet, nil
+		}
```

Comment:
How about moving this part into a separate function like `handleDroppingForRange(idx)` or something like this?